app/policies/post_policy.rb

```diff
@@ -1,5 +1,20 @@
 class PostPolicy < ApplicationPolicy
+  class Scope < Scope
+    def resolve
+      if user.present?
+        if user.admin? || user.moderator?
+          scope.all
+        else  
+          scope.where(:user_id => user.id)
+        end
+      else
+          scope.where(:user_id => nil)
```

Comment:
In this case, you want no posts returns.  The code you wrote here will return any `post` not assigned to a user (ie `post.user_id` is `nil`).  This may be equivalent to no posts, but may not be.